Refactor code to allow custom Blockly panels in Java plugins

Goldorion opened the pull request:
In #2239, I changed a lot of things to optimize the creation of new Blockly panels. However, it still didn't allow the creation of new custom Blockly panels inside Java plugins. This PR aims to change this by refactoring everything else necessary.

### Important points:
- `BlocklyEditorType` has been changed to a public record, so Java plugins can use the `BlocklyEditorToolbar` class.
- Changed `BlocklyLoader` to allow custom Blockly folders inside plugins
- Once a new block loader entry has been added inside `BlocklyLoader`, everything is automatized for them.
- Toolbox XML files have been moved to plugins (loaded inside `BlocklyJavaScriptsLoader` and renamed to `BlocklySpecialFilesLoader `)
- Added an event for Java plugins to allow them adding new sections to the `PreferencesDialog`
- Added 2 methods inside `PreferencesDialog` to add custom templates sections
- Added some Java documentation

### Changelog entries:
- Java plugins can now declare and use custom Blockly editors.
- Custom template sections for the preferences dialog can now be made with Java plugins.

Note: Normal sections are also theoretically possible, but without using a custom preference system or until #3071 is merged, it's not really possible.

Defeatomizer commented on src/main/java/net/mcreator/blockly/data/BlocklyLoader.java:
```diff
@@ -29,6 +31,20 @@ public static void init() {
 		INSTANCE = new BlocklyLoader();
 	}
 
+	private static final List<String> builtinCategories = new ArrayList<>() {{
```
Maybe pass all these categories within `Arrays.asList()` or `List.of()` as a constructor argument or add them inside the `BlocklyLoader()` constructor? Lists are not as complex as maps so I don't think they require things like this to be a separate block.

Goldorion replied:
I originally added them inside the constructor, but it caused an error in External Block Loader as the constructor wasn't finished. I'll go with the List.of option.

KlemenDEV reviewed (commented):
Maybe it would be better to make `blockly/js` and `blockly/toolbox` folders for more clarity?

Right now blockly JS and XML files are in the same folder and it can get a bit messy

KlemenDEV commented on src/main/java/net/mcreator/ui/init/BlocklySpecialFilesLoader.java:
```diff
@@ -19,37 +19,48 @@
 import java.util.regex.Pattern;
 
-public class BlocklyJavaScriptsLoader {
+public class BlocklySpecialFilesLoader {
```
I would also keep BlocklyJavaScriptsLoader and toolbox loader separate as they are different things, together with different folders

KlemenDEV commented:
On hold until I merge #2528, so this PR can then also add this panel to the system for complete coverage

Goldorion commented:
> On hold until I merge #2528, so this PR can then also add this panel to the system for complete coverage

😔 I just hope we will have enough time in 2022.4/2023.1 as the next update will surely be for the next MC update, so it would mean it's not merged before several months (and my plugin is not updated for several months).

KlemenDEV commented:
#2551 an older PR so it has priority over this one. Both are big PRs so this automatically means smaller priority, but rank about the same in size so age is deciding factor here.

Goldorion commented:
Oh I know and I understand. I just want to be able to update my plugin (because users are waiting after me for a few weeks now).

KlemenDEV commented:
Don't worry, #2551 is almost fully reviewed, I just need to review tests and do ingame tests and it is ready :)

Goldorion commented:
> Don't worry, #2551 is almost fully reviewed, I just need to review tests and do ingame tests and it is ready :)

👀

KlemenDEV reviewed (commented):
There are changes in this PR that do not seem related to the PR, for example `gui_msg_button.java.ftl`

Goldorion commented:
> There are changes in this PR that do not seem related to the PR, for example `gui_msg_button.java.ftl`

Strange... I didn't even touch to them when merging 🤔

Defeatomizer commented on src/main/java/net/mcreator/blockly/data/BlocklyLoader.java:
```diff
@@ -26,41 +31,53 @@ public static void init() {
+	 * @param name The category's name as written inside the XML file (e.g. <i>&lt;custom-thisName/&gt;</i>).
+	 */
+	public static void addBuiltinCategory(String name) {
+		builtinCategories.add(name);
```
If you want this to work, you need to wrap current variable initializer with `new ArrayList<>()` (`Arrays.asList()` and `List.of()` alone give you a list of fixed size so this code would throw an UOE).

KlemenDEV commented on plugins/generator-1.18.2/forge-1.18.2/templates/gui/gui_msg_button.java.ftl:
```diff
@@ -80,13 +80,15 @@ package ${package}.network;
 
 		<#assign btid = 0>
-        <#list data.getComponentsOfType("Button") as component>
+        <#list data.components as component>
```
This and other files still seem unrelated to this PR to me

Goldorion replied:
Again? I removed them 😞

Defeatomizer replied:
The merge commit resurrected these changes...

Goldorion replied:
yeah, because I reverted the commit that added those changes and when I checked the changed files again, they were gone...

Goldorion replied:
It should be good now. I copied the files from the repository (on from the website) and they are removed.
And here is my proof if it happens again 
![image](https://user-images.githubusercontent.com/38427877/208974595-5e52e46d-ce6f-449d-a248-76f0bf17602e.png)

KlemenDEV commented on src/main/java/net/mcreator/element/types/Overlay.java:
```diff
@@ -58,10 +58,6 @@ public Overlay(ModElement element) {
 		this.overlayTarget = "Ingame";
 	}
 
-	public boolean hasTextures() {
```
More unrelated changes, also seems to cause tests to fail. Please recheck PR in github "Files changed" tab

Defeatomizer replied:
You should also sync changes in WYSIWYG dialogs, they are the same thing.

Goldorion replied:
Weird... I have it locally and everything has been pushed to my branch 
![image](https://user-images.githubusercontent.com/38427877/208976736-0946598b-a414-4027-ab5a-eac9a84cb0b3.png)

Goldorion replied:
I checked everything and this weird behaviour is, in theory, from the past. Everything I saw was things I changed myself for this PR.

KlemenDEV commented on src/main/java/net/mcreator/blockly/data/BlocklyLoader.java:
```diff
@@ -26,41 +28,66 @@ public static void init() {
 
-	public ExternalBlockLoader getJSONTriggerLoader() {
-		return jsonTriggerLoader;
+	public ExternalBlockLoader getSpecificBlockLoader(String name) {
```
I think you can leave out specific here

Goldorion replied:
To allow custom Blockly panels in custom mod elements, we still need this new method, so Java plugins can also implement their own Blockly panels (which is the goal of this PR).

Defeatomizer replied:
I suppose Klemen meant the word `Specific` in the method name.

Goldorion replied:
> I suppose Klemen meant the word `Specific` in the method name.

Oh ok, my bad.

KlemenDEV replied:
Yep, just meant the name

Goldorion replied:
I used Specific to make it more different with the global getter returning all block loaders (and it's called getBlockLoaders)

KlemenDEV replied:
Maybe call the method `getAllBlockLoaders` and then this one `getBlockLoader` then

KlemenDEV commented on src/main/java/net/mcreator/blockly/data/BlocklyLoader.java:
```diff
@@ -26,41 +28,66 @@ public static void init() {
 
-	public ExternalBlockLoader getJSONTriggerLoader() {
-		return jsonTriggerLoader;
+	public ExternalBlockLoader getBlockLoader(String name) {
```
I think loading block loader should be done by BlocklyEditorType for key, not hardcoded string literal

Goldorion replied:
I will need to duplicate all methods because global triggers use the same system, but they can't use a BlocklyEditorType as they are not blockly blocks.

KlemenDEV replied:
Not sure what you mean, global triggers call `getBlockLoader` method?

Goldorion replied:
JSON triggers, sorry. However, the problem is still the same. They use the same system as the Java Blockly editors and so, using a `BlocklyEditorType` for methods will require to duplicate methods (to support json triggers) or to change this system, which will make the PR much bigger (or require a second PR).

KlemenDEV replied:
I see, however, I really don't like using hardcoded literals for this. I will think about it a bit and propose a solution, or alternatively we will use hardcoded strings for now

KlemenDEV replied:
You could add new BlocklyEditorType for JSON triggers, and make parameters that are not used null, which would indicate those features are not supported by this BlocklyEditorType

KlemenDEV commented on src/main/java/net/mcreator/generator/GeneratorStats.java:
```diff
@@ -185,6 +135,30 @@ public class GeneratorStats {
+		return generatorBlocklyBlocks;
+	}
+
+	public Set<String> getBlocklyBlocks(String name) {
```
This should also probably have BlocklyEditorType as argument, not literal

KlemenDEV reviewed (changes requested):
There may be also some other locations where using object instead of string literal may be more appropriate

Goldorion commented:
I made the requested changes (about using `BlocklyEditorType` (BET) instead of literal strings). I added 2 new BET to make everything work correctly, one for advancement blocks (`jsontriggers`) and a second one for global triggers (it was required in the end).

In addition to this, I also tried to change every place that still used a literal String to use the value from their BET object. I might have forgotten a few places (mostly for the starter block), but I checked with the Find in files option of Intellij and it seems everything was gone.

Finally, I added a new event for Java plugins to allow them to add new sections inside the `PreferencesDialog`, including sections for templates (added a general method + a specific method for BET). I also changed the translation key of Blockly templates (to use their BET translation key) and uniformize everything. The template sections are also a bit reorganized.

KlemenDEV commented:
> Edit: I just realized I only needed a simple loop to add a custom template section for each BET, so I made the change. However, is the new event still relevant or should I move it to https://github.com/MCreator/MCreator/pull/3072?

If #3072 will change this, you could leave it for this PR, but you can also keep it here for in-between if you want. I will be able to provide a better answer once I take a look at the PR fully

Goldorion commented:
> If https://github.com/MCreator/MCreator/pull/3072 will change this, you could leave it for this PR, but you can also keep it here for in-between if you want. I will be able to provide a better answer once I take a look at the PR fully

#3072 is mainly about changing how preferences are made to allow custom preferences entries that are saved inside a file. So I didn't touch on what is related to custom templates and their categories. This is why I'm not sure where it would fit best. It can fit in both PRs depending on how you see the thing.

KlemenDEV commented:
You can leave it in this PR then I would say

KlemenDEV commented on src/main/java/net/mcreator/blockly/java/BlocklyToProcedure.java:
```diff
@@ -52,7 +52,7 @@ public BlocklyToProcedure(Workspace workspace, ModElement parent, String sourceX
 		if (doc != null) {
 			// first we load data from startblock
-			Element trigger = XMLUtil.getFirstChildrenWithName(BlocklyBlockUtil.getStartBlock(doc, "event_trigger"),
+			Element trigger = XMLUtil.getFirstChildrenWithName(BlocklyBlockUtil.getStartBlock(doc, BlocklyEditorType.PROCEDURE.startBlockName()),
```
Couldn't you use `editorType.startBlockName()` here?

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyEditorType.java:
```diff
@@ -19,30 +19,13 @@
 package net.mcreator.ui.blockly;
 
-public enum BlocklyEditorType {
+public record BlocklyEditorType(String folder, String translationKey, String extension, String startBlockName) {
```
I suggest merging renaming `folder` to `registryName` and using same name for `translationKey` too, so removing `translationKey` argument here

Goldorion replied:
I thought to that, but I thought it was better to keep as many translations as possible.

KlemenDEV commented on src/main/java/net/mcreator/blockly/datapack/BlocklyToJSONTrigger.java:
```diff
@@ -48,6 +49,7 @@
 public class BlocklyToJSONTrigger extends BlocklyToCode {
 
 	private static final Logger LOG = LogManager.getLogger("Blockly2JSONTrigger");
+	private static final BlocklyEditorType editorType = BlocklyEditorType.JSON_TRIGGER;
```
`editorType` is argument of BlocklyToCode so redifining it here as static shadows that variable. Use property of object instead

Goldorion replied:
`editorType` is an argument of BlocklyToJava only. It was never implemented inside BlocklyToCode as it meant adding support for JSON triggers (and I wanted to avoid this at the beginning). However, now that they use it, I generalize it argument to BlocklyToCode.

KlemenDEV commented on src/main/java/net/mcreator/ui/blockly/BlocklyEditorType.java:
```diff
@@ -19,30 +19,13 @@
+	public static final BlocklyEditorType COMMAND_ARG = new BlocklyEditorType("cmdargs", "cmdtpl", "args_start");
+	public static final BlocklyEditorType FEATURE = new BlocklyEditorType("features", "ftpl", "feature_container");
+	public static final BlocklyEditorType JSON_TRIGGER = new BlocklyEditorType("jsontriggers", null, "advancement_trigger");
+	public static final BlocklyEditorType GLOBAL_TRIGGER = new BlocklyEditorType("triggers", null, null);
```
Global trigger is not really a blockly editor type so I would still keep string from this one hardcoded

It will be confusing for new code readers to see blockly editor types and see global trigger here where there is no blockly editor for global triggers

Goldorion replied:
> and a second one for global triggers (it was required in the end).

I explained why I added it in my long message above. So except if you want an even bigger (and messier) PR to accept both literal Strings and BlocklyEditorType objects, you need to choose between literal Strings for everything or BET objects for everything, including global triggers. However, I can try to change this behaviour in another PR after this one is merged.

Another simple solution might be to simply rename the class to something more global.

KlemenDEV replied:
The only location where it is called is here:

![image](https://user-images.githubusercontent.com/16374228/210326831-5849199f-0777-4ba2-b86b-b409aa25e2d3.png)

I see no reason why here you couldn't use hardcoded literal. Trigger loader does not even have to do anything with this PR as this pr addresses Blockly editors, not a system for loading triggers.

KlemenDEV replied:
The other case is addBlocklyFolder, which is also wrong to be used for triggers, and you have special if there:

![image](https://user-images.githubusercontent.com/16374228/210327030-f564f162-f537-46ad-8736-a19c4f55bd54.png)

So trigger stats could and should be handled by another method and thus eliminating the need for using GLOBAL_TRIGGER  as Blockly editor type

Goldorion replied:
Ok fine. However, its registry name will be hardcoded everywhere.

Defeatomizer commented on src/main/java/net/mcreator/generator/GeneratorStats.java:
```diff
@@ -147,24 +148,36 @@ public void addBlocklyFolder(GeneratorConfiguration genConfig, BlocklyEditorType
-		}
+
+		coverageInfo.put(type.registryName(), Math.min(
+				(((double) blocks.size()) / BlocklyLoader.INSTANCE.getExternalTriggerLoader().getExternalTrigers()
+						.size()) * 100, 100));
```
```suggestion
				(((double) blocks.size()) / BlocklyLoader.INSTANCE.getBlockLoader(type).getDefinedBlocks().size())
						* 100, 100));
```

Goldorion replied:
😬

KlemenDEV commented on src/main/java/net/mcreator/blockly/data/BlocklyLoader.java:
```diff
@@ -26,41 +33,67 @@ public static void init() {
+		add("text");
+		add("time");
+		add("advanced");
+		add("action");
```
This was called actions before, not action. Are you sure this is right?

Goldorion replied:
I surely removed the `s` by accident and I never noticed it.

KlemenDEV commented on src/main/java/net/mcreator/ui/dialogs/preferences/PreferencesDialog.java:
```diff
@@ -173,18 +177,24 @@ private void loadSections() {
-		new EditTemplatesPanel(this, L10N.t("dialog.preferences.page_armor_templates"), "templates/textures/armormaker",
-				"png");
+
+		addEditTemplatesPanel("ui_backgrounds", "backgrounds", "png");
```
The background is not in the templates folder, thus rendering the background useless and not working. Please, go through your changes when you do them and test them.

I can imagine seeing how changing this will work, but one should never assume and test all affected components.

KlemenDEV replied:
Also test if other templates still work

Goldorion replied:
I tested the 2 others, but not this one 🤦😓

KlemenDEV commented:
Ready for review?

Goldorion commented:
> Ready for review?

Oh yes, I was waiting for tests.

KlemenDEV commented:
Something is telling me coverage stats was not tested at all as even if removing blocks, it still shows 100% in stats which is ofc wrong..

KlemenDEV commented:
I fixed it, but please, do more tests on your PRs.

Otherwise, I have to question every single line when reviewing and test everything myself again, which in turn takes more time than just making PR from scratch.

Goldorion commented:
> Something is telling me coverage stats was not tested at all as even if removing block templates (for practically any type), it still shows 100% in stats which is ofc wrong..

I'm not sure to see what was the problem as coverage bar doesn't always show 100%. I removed some code templates for procedures, aitasks, cmdargs, features, jsontriggers and triggers and they all worked fine. (Your commit is still not on my computer)
![image](https://user-images.githubusercontent.com/38427877/210629790-c2911c19-1bf8-440c-9471-61953b9f2427.png)

KlemenDEV commented:
If you delete single one, it stays on 100% as all _ templates are counted resulting at more divided by less, which is more than 1, which is multiplied by 100 and capped at 100, showing 100 when there is not 100%.

Goldorion commented:
> If you delete single one, it stays on 100% as all _ templates are counted resulting at more divided by less, which is more than 1, which is multiplied by 100 and capped at 100, showing 100 when there is not 100%.
> 
> This is also clear on my commit: [12c4ee6](https://github.com/MCreator/MCreator/commit/12c4ee6c3581e47c3dc9557fbc015a103e12fdd2)
> 
> and looking at changed files, where you can see there were _ checks, but you then removed them

Oh ok, I understand. I always tested by deleting more than one element. I will be more careful from now on.

KlemenDEV commented:
> Oh ok, I understand. I always tested by deleting more than one element, explaining why I never noticed this problem. I will be more careful from now on.

The thing is coverage at 100% is the most important as otherwise, we may port all but say 2 templates but stats will show 100% and we would consider procedures ported to the new MC version, where in fact we would forget 2. Tests would not catch that either.

KlemenDEV reviewed (approved):
LGTM
